books/RayTracingTheNextWeek.html

```diff
@@ -2057,105 +2057,147 @@
+            if (!data) return magenta;
 
-            const auto color_scale = 1.0 / 255.0;
-            auto pixel = data + j*bytes_per_scanline + i*bytes_per_pixel;
+            // Clamp pixel coordintes to [0,width-1] x [0,height-1]
+            x = (x < 0) ? 0 : (x < image_width)  ? x : (image_width  - 1);
+            y = (y < 0) ? 0 : (y < image_height) ? y : (image_height - 1);
 
-            return color(color_scale*pixel[0], color_scale*pixel[1], color_scale*pixel[2]);
+            return data + y*bytes_per_scanline + x*bytes_per_pixel;
         }
 
       private:
+        const int bytes_per_pixel = 3;
         unsigned char *data;
-        int width, height;
+        int image_width, image_height;
         int bytes_per_scanline;
     };
 
+    // Restore MSVC compiler warnings
+    #ifdef _MSC_VER
+        #pragma warning (pop)
+    #endif
+
+    #endif
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-    [Listing [img-texture]: <kbd>[texture.h]</kbd> Image texture class]
+    [Listing [rtw_image]: <kbd>[rtw_stb_image.h] The rtw_image helper class]
+
+The preceding listing assumes that you have copied the `stb_image.h` header into a folder called
+`external`. Adjust according to your directory structure.
+
+If you are writing your implementation in a language other than C or C++, you'll need to locate (or
+write) an image loading library that provides similar functionality.
 
 <div class='together'>
-The representation of a packed array in that order is pretty standard. Thankfully, the [stb_image][]
-package makes that super simple -- just write a header called `rtw_stb_image.h` that also deals with
-some compiler warnings:
+The `image_texture` class uses the `rtw_image` class:
 
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
-    #ifndef RTWEEKEND_STB_IMAGE_H
-    #define RTWEEKEND_STB_IMAGE_H
+    #include "rtweekend.h"
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
+    #include "rtw_stb_image.h"
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    #include "perlin.h"
 
-    // Disable pedantic warnings for this external library.
-    #ifdef _MSC_VER
-        // Microsoft Visual C++ Compiler
-        #pragma warning (push, 0)
-    #endif
+    ...
 
-    #define STB_IMAGE_IMPLEMENTATION
-    #include "external/stb_image.h"
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
+    class image_texture : public texture {
+      public:
+        image_texture() {}
+        image_texture(const char* filename) : image(filename) {}
 
-    // Restore warning levels.
-    #ifdef _MSC_VER
-        // Microsoft Visual C++ Compiler
-        #pragma warning (pop)
-    #endif
+        virtual color value(double u, double v, const vec3& p) const override {
```

Comment:
<For prosperity>

remove virtual for override alone

Reply:
Preview of coming attractions.